skfuzzy/control/controlsystem.py

```diff
@@ -329,28 +132,266 @@ def inputs(self, input_dict):
-        fig = ControlSystemVisualizer(self).view()
-        fig.show()
+    def compute(self):
+        """
+        Compute the fuzzy system.
+        """
+        # TODO: Tracking and caching
+
+        # Check if any fuzzy variables lack input values and fuzzyfy inputs
+        for antecedent in self.ctrl.antecedents:
+            if antecedent.input[self] is None:
+                raise ValueError("All antecedents must have input values!")
+            if antecedent.terms.values()[0].membership_value[self] is not None:
+                raise RuntimeError("Antecedent already has calculated "
+                "membership.  Are you trying to computer a simulation multiple "
+                "times?  Create multiple ControlSystemSimulation objects "
+                "instead.")
+            CrispValueCalculator(antecedent, self).fuzz(antecedent.input[self])
+
+        # Calculate rules, taking inputs and accumulating outputs
+        for rule in self.ctrl.rules:
+            self.compute_rule(rule)
+
+
+        # Collect the results and present them as a dict
+        for consequent in self.ctrl.consequents:
+            consequent.output[self] = \
+                CrispValueCalculator(consequent, self).defuzz()
+            self.output[consequent.label] = consequent.output[self]
+
+    def compute_rule(self, rule):
+        """
+        Implements rule according to the three step method of
+        Mamdani inference: Aggregation, activation, and accumulation
+
+        """
+        # Step 1: Aggregation.  This finds the net accomplishment of the
+        #  antecedent by AND-ing or OR-ing together all the membership values
+        #  of the terms that make up the accomplishment condition.
+        #  The process of actually aggregating everything is delegated to the
+        #  FuzzyVariableTermAggregation class, but we can tell that class
+        #  what aggregation style this rule mandates
+
+        if isinstance(rule.antecedent, FuzzyVariableTermAggregate):
+            rule.antecedent.agg_method = rule.aggregation_method
+        rule.aggregate_firing[self] = rule.antecedent.membership_value[self]
+
+        # Step 2: Activation.  The degree of membership of the consequence
+        #  is determined by the degree of accomplishment of the antecedent,
+        #  which is what we determined in step 1.  The only difference would
+        #  be if the consequent has a weight, which we would apply now.
+        for c in rule.consequent:
+            assert isinstance(c, WeightedConsequent)
+            c.activation[self] = rule.aggregate_firing[self] * c.weight
+
+        # Step 3: Accumulation.  Apply the activation to each consequent,
+        #   accumulating multiple rule firings into a single membership value.
+        #   The process of actual accumulation is delegated to the
+        #   FuzzyVariableTerm which uses its parent's accumulation method
+        for c in rule.consequent:
+            assert isinstance(c, WeightedConsequent)
```

Comment:
assert statements help my IDE (pyCharm) understand the variable types that are being used, so I use them even when it's somewhat obvious what the variable type is.